src/replication.c

```diff
@@ -448,6 +449,245 @@ send_replica_handshake_query(replica_t *replica, spec_t *spec)
+		REPLICA_NOTICELOG("Failed to inform to mgmt_eventfd for replica(%p)\n", replica);
+		ret = -1;
+	}
+
+	return ret;
+}
+
+/*
+static bool
+any_ongoing_snapshot_command(spec_t *spec)
+{
+	replica_t *replica;
+	mgmt_cmd_t *cmd;
+	MTX_LOCK(&spec->rq_mtx);
+	TAILQ_FOREACH(replica, &spec->rq, r_next) {
+		MTX_LOCK(&replica->r_mtx);
+		TAILQ_FOREACH(cmd, &replica->mgmt_cmd_queue, mgmt_cmd_next) {
+			if (cmd->io_hdr->opcode == ZVOL_OPCODE_SNAP_CREATE) {
+				MTX_UNLOCK(&replica->r_mtx);
+				MTX_UNLOCK(&spec->rq_mtx);
+				return true;
+			}
+		}
+		MTX_UNLOCK(&replica->r_mtx);
+	}
+	MTX_UNLOCK(&spec->rq_mtx);
+	return false;
+}
+*/
+
+static int
+is_volume_healthy(spec_t *spec)
+{
+	if (spec->healthy_rcount != spec->replication_factor)
+		return false;
+	return true;
+}
+
+/*
+ * This function quiesces write IOs, waits for ongoing write IOs
+ * If volume is not healthy or timeout happens while waiting for ongoing IOs,
+ * write IOs will be allowed, and false will be returned.
+ * If there are no pending write IOs and volume is healthy, true will be returned.
+ * rq_mtx is required to be held by caller
+ */
+static bool
+pause_and_timed_wait_for_ongoing_ios(spec_t *spec, int sec)
+{
+	struct timespec last, now, diff;
+	bool ret = false;
+	bool write_io_found = false;
+	replica_t *replica;
+
+	spec->lu->quiesce = 1;
+
+	clock_gettime(CLOCK_MONOTONIC, &last);
+	timesdiff(last, now, diff);
+
+	while ((diff.tv_sec < sec) && (is_volume_healthy(spec) == true)) {
+		write_io_found = false;
+		TAILQ_FOREACH(replica, &spec->rq, r_next) {
+			if (replica->inflight_write_io_cnt != 0) {
+				write_io_found = true;
+				break;
+			}
+		}
+		if (write_io_found == false) {
+			if (spec->inflight_write_io_cnt != 0) {
+				write_io_found = true;
+				break;
+			}
+		}
+		if (write_io_found == false) {
+			ret = true;
+			break;
+		}
+		MTX_UNLOCK(&spec->rq_mtx);
+		sleep (1);
+		MTX_LOCK(&spec->rq_mtx);
+		timesdiff(last, now, diff);
+	}
+
+	if (ret == false)
+		spec->lu->quiesce = 0;
+
+	return ret;
+}
+
+int istgt_lu_destroy_snapshot(spec_t *spec, char *snapname)
+{
+	replica_t *replica;
+	TAILQ_FOREACH(replica, &spec->rq, r_next)
+		send_replica_snapshot(spec, replica, snapname, ZVOL_OPCODE_SNAP_DESTROY, NULL);
+	return true;
+}
+
+/*
+ * This API will create snapshot with given name on the spec.
+ * It will wait for io_wait_time seconds to complete ongoing IOs.
+ * Overall, this API will wait for wait_time seconds to get response
+ * for snapshot command (this includes io_wait_time).
+ * In case of any failures, snapshot destroy command will be sent to all replicas.
+ */
+int istgt_lu_create_snapshot(spec_t *spec, char *snapname, int io_wait_time, int wait_time)
+{
+	bool r;
+	replica_t *replica;
+	int free_rcomm_mgmt = 0;
+	struct timespec last, now, diff;
+	struct rcommon_mgmt_cmd *rcomm_mgmt;
+	int rc;
+
+	clock_gettime(CLOCK_MONOTONIC, &last);
+	MTX_LOCK(&spec->rq_mtx);
+
+	if (is_volume_healthy(spec) == false) {
```

Comment:
You can move volume healthy check in func pause_and_timed_wait_for_ongoing_ios and so does MTX_LOCK. At line 587, we are unlocking it without locking it in the same scope which is little difficult to maintain going forward.

Reply:
pause_and_timed_.. function does the check for volume health. But, before that, it quiesce IOs also, which is not required if volume is not healthy.
Regarding line 587, comment is available for the function to have lock before calling this function.